helix-view/src/document.rs

```diff
@@ -648,12 +704,32 @@ impl Document {
 
         if !transaction.changes().is_empty() {
             self.version += 1;
+            let reverted_tx = transaction.invert(&old_doc);
+
+            // Consider H to be the text of the file in git HEAD, and B₁ to be
+            // the text when buffer initially loads. H → B₁ is the changeset
+            // that describes the diff between HEAD and buffer text. Inverting
+            // this produces B₁ → H, which is initially saved to `diff_changes`.
+            // In the next edit, buffer text changes to B₂. The transaction
+            // describes the change B₁ → B₂. Inverting this gives B₂ → B₁.
+            // Composing this with the saved `diff_changes` gives us
+            // (B₂ → B₁) → (B₁ → H) = B₂ → H. Whenever examining a change X₁ → X₂,
+            // we need to know the contents of the text at state X₁ to know where
+            // to apply the operations in the changeset. The advantage of inverting and
+            // composing this way instead of simply composing (which would give
+            // us H → B₂ in this example) is that we no longer need the HEAD text
+            // and we can instead use the current text in the buffer.
+            if let Some(changes) = self.diff_changes.take() {
+                let reverted_changes = reverted_tx.changes().clone();
+                let changes = reverted_changes.compose(changes);
+                self.diff_changes = Some(changes);
+                self.diff_with_base();
+            }
```

Comment:
We use changesets now, and use the approach described in the comment. It also has the advantage that we don't have to manage a cache of the file contents in HEAD since we don't need it to generate the diff.

This works fine for most of the cases except for deleting something and undoing it (there are some other rough corners, but they can be easily solved at the rendering level). Normally, composing a change and it's undo operation produces an empty changeset. But since we use inverted changes here, it doesn't apply cleanly.

I'll write a better explanation of the problem tomorrow, but it boils down to the fact that composing for example `[Delete(5)]` and `[Insert("hello")]` (note the order) does not produce an empty changeset, where as composing an insert and _then_ a delete gives a changeset of `[]`.

Relevant test case:
<details><summary>Patch for the test</summary>

```diff
:100644 100644 daf4a77e 00000000 M	helix-core/src/transaction.rs

diff --git a/helix-core/src/transaction.rs b/helix-core/src/transaction.rs
index daf4a77e..9aed26e8 100644
--- a/helix-core/src/transaction.rs
+++ b/helix-core/src/transaction.rs
@@ -579,6 +579,17 @@ mod test {
     use super::*;
     use crate::State;
 
+    #[test]
+    fn composing_delete_and_insert_leaves_empty_changeset() {
+        let rope = Rope::from("");
+        let mut delete = ChangeSet::new(&rope);
+        delete.delete(5);
+        let mut insert = ChangeSet::new(&rope);
+        insert.insert("hello".into());
+
+        let changes = delete.compose(insert);
+        assert_eq!(changes.changes, &[]);
+    }
     #[test]
     fn composition() {
         use Operation::*;
```
</details>

```rust
fn composing_delete_and_insert_leaves_empty_changeset() {
    let rope = Rope::from("");
    let mut delete = ChangeSet::new(&rope);
    delete.delete(5);
    let mut insert = ChangeSet::new(&rope);
    insert.insert("hello".into());

    let changes = delete.compose(insert);
    assert_eq!(changes.changes, &[]);
}
```